lib/model.js

```diff
@@ -3937,18 +3937,24 @@ class Model {
     // Clone values so it doesn't get modified for caller scope and ignore undefined values
     values = _.omitBy(values, value => value === undefined);
 
-    const changedBefore = this.changed() || [];
-
     options = options || {};
     if (Array.isArray(options)) options = {fields: options};
 
     options = Utils.cloneDeep(options);
     const setOptions = Utils.cloneDeep(options);
     setOptions.attributes = options.fields;
+
+    // We want to keep track of the fields changed via side effects
+    const pushed_changed = this._changed;
+    this._changed = {};
     this.set(values, setOptions);
 
     // Now we need to figure out which fields were actually affected by the setter.
-    const sideEffects = _.without.apply(this, [this.changed() || []].concat(changedBefore));
+    const sideEffects = this.changed();
```

Comment:
? I don't know what you mean by "Why don't we send new computed `_changed` to `sideEffects`?" Isn't that what I am doing?

By "Modifying `_changed` directly feels dirty" are you referring this line: "`this._changed = {};`"? I agree that's a bit dirty however I need to reset `this._changed` to an empty set. When `this.set(values, setOptions)` runs it populate the `this._changed`. The problem is `_.without.apply(this, [this.changed() || []].concat(changedBefore));` (which is is really conceptually `this.changed() - the elements in changedBefore`) does the wrong thing. If a key has been `changedBefore` but also modified by `this.set(values, setOptions)` (ie: the user's call to `someDbRecord.update(...);`) it won't get written to the DB. Essentially:

```
// someDbRecord is created with key: "Z" which is a virtual key mapping to "virtualKey"
someDbRecord.key = "A"; // this populates "changedBefore" with "virtualKey" during the update()
await someDbRecord.update({ key: "B" }); // does nothing because it's virtualKey is in changedBefore

someDbRecord.key === "B"
await reload();
someDbRecord.key === "Z" // ! argh update did nothing
```

So the fix is to figure out what changed before, push the list of changedKeys, blank/clear the list of this._changed keys, figure out what changed for the update (that's the side effect list).. The next step is to pop the list of changed keys, which we do by appending the sideEffectFields with the original changedBefore keys (making sure to deduplicate keys). 

I don't know how to do this without blanking the changed list like this: `this._changed = {}` and restoring the changed list like this: this `._changed = _.union(...);`